[compiler v2] Copy Propagation Optimization

vineethk opened the pull request:
This PR adds the following analyses and transformations on the stackless bytecode, and enables them by default in the compiler v2:
* Available copies analysis
* Copy propagation transformation
* Dead store elimination

The overall effect is (potentially): (a) reduction in the number of locals created, (b) reduction in the number of instructions.

~This PR also adds the "uninitialized use checker" for compiler v2, which checks the use of uninitialized locals (also performed on the stackless bytecode). It performs initialization state analysis to do these checks. This is enabled in compiler v2 as long as safety checks are enabled (which they are, unless explicitly disabled).~

### Description

#### Available copies analysis

This analysis computes for each program point, which copies of the form `a := b` (`a` assigned to `b`) are definitely available. It is implemented as a forwards must analysis.

#### Copy propagation transformation

This analysis transforms the bytecode by making use of the available copies information, and replacing uses of locals with their copy-chain heads. This transformation creates a lot of dead stores.

For example:
```
let a = 1;
let b = a;
let c = b;
c  // copy-chain: a <- b <- c
```

is transformed into:
```
let a = 1;
let b = a;  // dead store
let c = a;  // dead store, copy-chain head of `b` is `a`
a  // copy-chain head of `c` is `a`
```

#### Dead store elimination

This transformation deletes stackless bytecode instructions that do wasteful writes to locals:
- write to a local that is never used again
- self assignment to a local (i.e., `x := x`)

#### ~Initialization state analysis~

~This analysis computes for each local at each program point, is it:~
- ~definitely initialized~
- ~definitely not initialized~
- ~maybe initialized~

~It is implemented as a "forwards may" analysis. These analysis results are used by "uninitialized use checker" to see if any uses of locals are not "definitely initialized".~


### Test Plan
* All the transactional tests comparing v1 vs v2 pass as before, except `deep_exp` which now passes correctly with the same o/p in v1 and v2.
* Several copy propagation specific tests added with IR displayed for relevant stages of the optimization pipeline.
* ~Ported uninitialized use check related tests from compiler v1 tests as well as added some new ones.~

Closes https://github.com/aptos-labs/aptos-core/issues/11276

vineethk commented on third_party/move/move-compiler-v2/tests/uninit-use-checker/use_before_assign.exp:
```diff
@@ -0,0 +1,29 @@
+error: use of unassigned local `x`
+  ┌─ tests/uninit-use-checker/use_before_assign.move:4:9
+  │
+4 │     let y = x;
```
This is a bug due to location information not properly propagated to stackless bytecode. I have filed a bug for it: https://github.com/aptos-labs/aptos-core/issues/11599

vineethk commented:
@brianrmurphy suggested two tests of this form:
```
{
    let a = unknown();
    let b = a;
    let c = b;

    b = other();

    let t = a == c;
    test t;
}

{
    let a = unknown();
    let b = a;
    let c = b;

    a = other();

    let t = b == c;
    test t;
}
```

I am adding this comment to track addition of these tests.

brmataptos commented on third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs:
```diff
@@ -0,0 +1,348 @@
+    /// Note that it is a required invariant that the copy chain is acyclic, else we panic.
+    /// The natural way of constructing the copy chain for move bytecode (like in this file) ensures this.
+    pub fn get_head_of_copy_chain(&self, mut tmp: TempIndex) -> TempIndex {
+        let mut visited = BTreeSet::from([tmp]);
```
This seems like an expensive way of checking, especially since you expect to never happen.  How about a counter instead?
```
   let mut counter = 0;
   let usize max_counter = self.0.len();
   while let Some(src) = self.0.get(&tmp) {
      counter += 1;
      if counter >= max_counter {  panic ... }
      tmp = *src;
   }
   tmp

```

vineethk replied:
Nice idea.

What do you think about removing this checking related code completely? I added it initially as a "debug" code to make sure I did not have bugs in the way I construct copy chains, but I think based on how they are constructed, we can prove that there will be no cycles.

wrwg replied:
I think we should keep this kind of checks, they are also good for documentation. As to serve as doc/spec, the current code is a better readable than Brian's... Performance probably not that important, these chains are not that deep. But both are good with me.

vineethk commented on third_party/move/move-compiler-v2/tests/testsuite.rs:
```diff
@@ -303,10 +350,16 @@ impl TestConfig {
                         Self::check_diags(out, &env);
-                        if self.dump_annotated_targets {
+                        if self.dump_annotated_targets
+                            && (self.dump_for_only_some_stages.is_none() // dump all stages
```
I briefly considered that, but you can run the same analysis multiple times in a pipeline. Using string-based naming (at least based on naming the analysis/transformation/check) is less expressive.

vineethk commented on third_party/move/move-compiler-v2/tests/testsuite.rs:
```diff
@@ -303,10 +350,16 @@ impl TestConfig {
+                                || self
+                                    .dump_for_only_some_stages
+                                    .as_ref()
+                                    .is_some_and(|list| list.contains(&(i - 1))))
```
I'll add that info for the `run_with_hook` docs.

vineethk commented on third_party/move/move-compiler-v2/tests/testsuite.rs:
```diff
@@ -44,6 +45,11 @@ struct TestConfig {
     generate_file_format: bool,
     /// Whether we should dump annotated targets for each stage of the pipeline.
     dump_annotated_targets: bool,
+    /// Optionally, dump annotated targets for only certain stages of the pipeline.
```
This contradicts your "Indexing starting at 1? Who does that?" comment :)

Joking aside, I will add the comment. I did actually have this in comments, and then thought, this is too obvious, and removed it. But in light of `run_with_hook` counting pipeline stages with 1, it is better to include the comment.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs:
```diff
@@ -0,0 +1,381 @@
+    Some(s)
+}
+
+/// Format a vector of `locals`.
```
It it just a private helper used in other formatters in this file, to reduce code duplication.

Because these are used for annotation formatting, which can be specific to the analyses, I am not sure if this specific style of formatting is more generally useful and thus belongs as a public method in the `FunctionTarget` interface.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs:
```diff
@@ -0,0 +1,381 @@
+    ///            +--------+
+    ///            | bottom |
+    ///            +--------+
+    /// Note that bottom is not explicitly represented in the code.
```
Maybe I could be direct about what "bottom" represents here: the reason why we get `Option<Initialized>` at a program point is because that program point is unreachable, and therefore, the dataflow analysis does not calculate anything at that program point. At such program points, we have `bottom` values.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/dead_store_elimination.rs:
```diff
@@ -0,0 +1,89 @@
+    fn transform(code: Vec<Bytecode>, live_vars_annotation: &LiveVarAnnotation) -> Vec<Bytecode> {
+        let mut new_code = vec![];
+        for (offset, instr) in code.into_iter().enumerate() {
+            if let Bytecode::Assign(_, dst, ..) | Bytecode::Load(_, dst, _) = instr {
```
So, the order is usually: avail copy, copy propagation, live var, dead store elimination.

Thus, in the example you provided, we would have:

```
y = z;
// y is dead here onwards
x = z;  // copy propagation
// z is dead here onwards
```

Therefore, both assignments would be eliminated.

However, liveness is not transitive (that is, it does not check if all downstream uses are being used to write to variables that will be used).

So, if you have:
```
b = true;
b = !b;  //  `!` can never abort
// b is dead
```

You cannot eliminate the these stores to `b`. To do this, you would need a "neededness analysis": a backwards dataflow analysis that computes needed statements expressions. I initially set out to do this, but realized that due to most operations (like addition) possibly aborting, almost everything becomes needed, and you can get very similar mileage from this simpler dead store elimination (which will automatically do more code reduction once I land const prop/folding).

wrwg replied:
Copy propagation is one simple way to deal with transitive liveness. You just skip all these intermediates in the transitive chain, and then the dead assignments are exposed. Its a standard procedure.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/copy_propagation.rs:
```diff
@@ -0,0 +1,86 @@
+        let new_code = Self::transform(&target, code, avail_copies);
+        data.code = new_code;
+        // Annotations may no longer be valid after this transformation, so remove them.
+        data.annotations.clear();
```
The annotations we add are associating code offsets to information computed at these code offsets. A transformation can change these code offsets, so we should conservatively assume that annotations are no longer valid. Unless we add some abstraction for annotations that are independent of code offsets (which there are none, as far as I know), we would not know which ones to clear.

I think we should do this for every transformation pass (i.e., ones that change code), but I have not done it in this PR.

wrwg replied:
This is a good point, we have not handled this explicitly until now. Lets keep what Vineeth has here for this PR, but come back to this problem. One thing I can imagine is we delegate it to the pipeline code. Perhaps the trait has some `fn invalidates_code_offsets(&self): bool` and then it happens automatically.

fEst1ck commented on third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs:
```diff
@@ -0,0 +1,381 @@
+
+/// State of initialization of a local at a given program point.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Initialized {
```
Perhaps we can we implement some common constructions on lattice. In this case, a general construction could be given a set T, attach a top element and a bottome element to it, such that for all t in T, bot <= t, t <= top, and nothing else. Another common one might be lattice products, which perhaps already appeared in the code base, where an `AbstractDomain` has other `AbstractDomain` fields and the join is fieldwise join.

fEst1ck replied:
https://github.com/aptos-labs/aptos-core/issues/11632

vineethk replied:
Agree, see comment reply on a similar comment below.

fEst1ck commented on third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs:
```diff
@@ -0,0 +1,381 @@
+
+impl AbstractDomain for InitializedState {
+    /// `join` of two initialized states is the point-wise join for each local.
+    /// The result is `JoinResult::Changed` if any of the local joins causes a change.
```
This seems also a very common situation, where we each local has an abstract value from a lattice T, and the abstract state of a program point is T^n where the program n locals. In code: `impl<T: AbstracDomain> AbstrctDomain for Vec<T>`

fEst1ck replied:
https://github.com/aptos-labs/aptos-core/issues/11632

vineethk replied:
Note that I am using a `im::Vector` here, but your general point still stands. I think it might be a good exercise to do some of these abstractions as a common refactoring pass. Given that this PR is already quite large, I think it makes sense to do this in a different pass: thank you for filing an issue to track this.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs:
```diff
@@ -0,0 +1,348 @@
+    fn join(&mut self, other: &Self) -> JoinResult {
+        let mut result = JoinResult::Unchanged;
+        let prev_copies = std::mem::take(&mut self.0);
+        for (dst_, src_) in &other.0 {
```
Perhaps `other_src` and `other_dst` is more verbose but better readable.

wrwg replied:
Rust has some pretty cool stuff for modifying maps efficiently. Not important here but in general you do:

```
match map.entry(*key) {
  Vacant(handle) => <<insert based on handle>>,
  Occupied(handle) => <<modify or delete based on handle>>
}
```

This way you could instead mutate `self.0` directly instead of reconstructing it.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs:
```diff
@@ -0,0 +1,348 @@
+                    self.0.insert(*dst_, *src_);
+                }
+            }
+            // else: a copy of the form (dst_, _) is not already available in self, so no change to self.
```
is not already --> was previously not

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    fn name(&self) -> String {
+        "available_copies".to_string()
```
Please check the other processors, but I think the names are currently caml-cased.

vineethk replied:
I may have named this based on how analyses under move-model/bytecode are named (eg, livevar_analysis.rs), however, the ones under compiler v2 pipeline are named using caml case.

Changing these for consistency.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/copy_propagation.rs:
```diff
@@ -0,0 +1,86 @@
+                .before(&(offset as CodeOffset))
+                .unwrap_or(&default_avail_copies);
+            let mut propagated_src = |dst| avail_copies.get_head_of_copy_chain(dst);
+            new_code.push(instr.remap_src_vars(target, &mut propagated_src));
```
Great that you found `remap_src_vars`, this is IIRC also used by the provers CP. We should double check that this is still correctly implemented but otherwise it should do exactly what you want.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/copy_propagation.rs:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    fn name(&self) -> String {
+        "copy_propagation".to_string()
```
Align with naming conventions.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/dead_store_elimination.rs:
```diff
@@ -0,0 +1,89 @@
+        // so we don't need to prune unused locals here for various fields of `data` (like `local_types`).
+        data.code = new_code;
+        // Annotations may no longer be valid after this transformation, so remove them.
+        data.annotations.clear();
```
This is OK for now, as discussed above.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs:
```diff
@@ -0,0 +1,381 @@
+// Copyright © Aptos Foundation
```
Is there a reason why this need to be in the same PR? Just asking.

vineethk replied:
No, I originally wanted to do stacked PRs but figured out that this cannot be done with forked repo model.

I am removing this checker from this PR and will create a separate PR for it.

wrwg reviewed (commented):
Looks great! Good example how to use the dataflow analysis framework.

I wonder whether we should factor safety checks out of this PR. The copy propagation is needed for us to pass some of the e2e tests (too many locals), but safety not.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs:
```diff
@@ -0,0 +1,348 @@
+            let old_src = self.0.insert(dst, src);
+            if let Some(old_src) = old_src {
+                panic!(
+                    "ICE: copy `$t{} = $t{}` already available, \
```
I have removed the "ICE", but kept the panic (as these are similar to panics in `reference_safety_processor`, eg).

vineethk commented:
@wrwg @brmataptos I have addressed all comments. I have also removed the uninitialized use checker from this PR as requested, I will file a separate PR for it.

PTAL.

wrwg commented on third_party/move/move-model/bytecode/src/stackless_bytecode.rs:
```diff
@@ -484,7 +484,37 @@ impl Bytecode {
+            Bytecode::SaveMem(_, _, _)
+            | Bytecode::SaveSpecVar(_, _, _)
+            | Bytecode::Prop(_, _, _) => {
+                vec![]
```
Have not seen last time. Thats kind of wrong semantics, so it would be more adequate to panic that this not implemented.

wrwg replied:
Also, those operations really do not appear in the compiler. They are all introduced as part of code instrumentation in the prover pipeline.

vineethk replied:
Done.

brmataptos commented on third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs:
```diff
@@ -0,0 +1,337 @@
+                    // We are removing the available copy (dst, src) from self.
+                    result = JoinResult::Changed;
+                } else {
+                    self.0.insert(*dst_, *src_);
```
Why do we need an insert here?

vineethk replied:
This means both self and other have this copy, so when we join, we keep that copy. I have added a comment to clarify.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs:
```diff
@@ -0,0 +1,337 @@
+    /// Make a copy `dst := src` available.
+    /// If either `dst` or `src` is in `borrowed_locals`, then the copy is not made available.
+    /// To call this method, `dst := x` should not already be available for any `x`.
+    fn make_copy_available(
```
I think that having these two separated out is nice from an abstraction stand point (and the pre-conditions of the function are documented and checked) and future code development stand point (if we decide to do more involved things with the borrowed variables). I am keeping these abstractions for now, instead of combining them.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs:
```diff
@@ -0,0 +1,337 @@
+    /// Kill all available copies of the form `x := y` where `x` or `y` is `tmp`.
+    /// If `tmp` is in `borrowed_locals`, then no copies are killed.
+    fn kill_copies_with(&mut self, tmp: TempIndex, borrowed_locals: &BTreeSet<TempIndex>) {
+        if !borrowed_locals.contains(&tmp) {
```
Fair enough, doing so.

vineethk commented:
@brmataptos I believe I have addressed all your comments. PTAL.

rahxephon89 reviewed (approved):
LGTM, thanks!

wrwg commented:
Could we land this? Right now ` BUILDING AptosTokenObjects  error: exceeded maximal local count: 255` blocks me from testing #11701 against e2e-move-tests.
